### PR TITLE
Feature/psyneu 46

### PR DESCRIPTION
**Pull request by emekauja**
Add and remove ports from metadata functionality
Screenshot 2023-06-19 at 9 20 23 PM

### CODE REVIEW

File: src/client/components/views/editView/mechanisms/AGTControlMechanism/MechMetadata.js [hunk 1]
```diff
@@ -34,6 +35,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/AGTControlMechanism/MechMetadata.js]: what does 'Composition 2' represents?

File: src/client/components/views/editView/mechanisms/AutoLearningMechanism/MechMetadata.js [hunk 2]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/AutoLearningMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ComparatorMechanism/MechMetadata.js [hunk 3]
```diff
@@ -31,6 +32,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ComparatorMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/CompositionInterfaceMechanism/MechMetadata.js [hunk 4]
```diff
@@ -32,6 +33,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/CompositionInterfaceMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ContrastiveMechanism/MechMetadata.js [hunk 5]
```diff
@@ -36,6 +37,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ContrastiveMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ControlMechanism/MechMetadata.js [hunk 6]
```diff
@@ -37,6 +38,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ControlMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/DDM/MechMetadata.js [hunk 7]
```diff
@@ -36,6 +37,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/DDM/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/DefaultProcessingMechanism/MechMetadata.js [hunk 8]
```diff
@@ -34,6 +35,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/DefaultProcessingMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/EpisodicMemoryMechanism/MechMetadata.js [hunk 9]
```diff
@@ -34,6 +35,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/EpisodicMemoryMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/GatingMechanism/MechMetadata.js [hunk 10]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/GatingMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/IntegratorMechanism/MechMetadata.js [hunk 11]
```diff
@@ -31,6 +32,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/IntegratorMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/KWTAMechanism/MechMetadata.js [hunk 12]
```diff
@@ -36,6 +37,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/KWTAMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/KohonenLearningMechanism/MechMetadata.js [hunk 13]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/KohonenLearningMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/KohonenMechanism/MechMetadata.js [hunk 14]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/KohonenMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/LCAMechanism/MechMetadata.js [hunk 15]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/LCAMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/LCControlMechanism/MechMetadata.js [hunk 16]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/LCControlMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/LearningMechanism/MechMetadata.js [hunk 17]
```diff
@@ -37,6 +38,20 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const memoisedOptions = React.useMemo(() => optionsValue, [optionsValue]);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/LearningMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ModulatoryMechanism/MechMetadata.js [hunk 18]
```diff
@@ -34,6 +35,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ModulatoryMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ObjectiveMechanism/MechMetadata.js [hunk 19]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ObjectiveMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/OptControlMechanism/MechMetadata.js [hunk 20]
```diff
@@ -36,6 +37,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/OptControlMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/PredictionErrorMechanism/MechMetadata.js [hunk 21]
```diff
@@ -34,6 +35,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/PredictionErrorMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/ProcessingMechanism/MechMetadata.js [hunk 22]
```diff
@@ -31,6 +32,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/ProcessingMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/RecurrentTransferMechanism/MechMetadata.js [hunk 23]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/RecurrentTransferMechanism/MechMetadata.js]: why are we repeating this function?

File: src/client/components/views/editView/mechanisms/TransferMechanism/MechMetadata.js [hunk 24]
```diff
@@ -35,6 +36,18 @@ function MechMetadata(props) {
 
   const [optionsValue, setOptions] = React.useState(() => options);
   const optionKeys = toObject(Object.entries(options));
+  const [value, setValue] = React.useState(() => ['Composition 2']);
+
+  const handleMenuValueChange = (id) => {
```
Comment (ddelpiano) [inline src/client/components/views/editView/mechanisms/TransferMechanism/MechMetadata.js]: why are we repeating this function?

**Review by ddelpiano — CHANGES_REQUESTED** [review]
I understand that copying and pasting code could look like a fast solution for a shared problem but it's not fun, mantainable and we need to avoid it.

**Comment by emekauja** [comment]
We no longer use any metadata.js just one mechMetadata.js in shared folder all this has been deleted in the PR @ddelpiano

**Review by ddelpiano — DISMISSED** [review]

**Pull request merged**